docs/dev/analysis/tbl-cell-borders.rst

```diff
@@ -0,0 +1,153 @@
+
+Table Cell Borders
+------------------
+
+In PowerPoint, a table cell can have borders on all 4 sides, plus two diagonal lines (bottom left to top right and
+top left to bottom right) for a total of 6 possible borders, each specified by a child element of the table cell
+properties element.
+
+Each border is further customized by a set of attributes and child elements common to all borders (``LineFormat``).
+
+In order to provide ``LineFormat`` functionality, each border ``_ln`` object is wrapped in a ``_CellBorderLine``,
+which implements the ``get_or_add_ln`` method, and can be then wrapped by ``LineFormat``.
+
+Protocol
+--------
+
+**Starting with a default cell** A new cell has no borders or style.
+New borders can be created by accessing the borders property::
+
+    >>> table = shapes.add_table(rows, columns, left, top, width, height).table
+    >>> cell = table.rows[0].cells[0]
+    >>> cell
+    <pptx.shapes.table._Cell object at 0x...>
+    >>> cell.borders
+    <pptx.shapes.table._CellBorders object at 0x...>
+
+**Create a border** A cell border is created once it's accessed for the first time.
+It won't be visible, however, until applying line width and color for its ``LineFormat``.
+This can be done with any one of the 6 cell borders::
+
+    >>> cell.borders.top
+    <pptx.dml.line.LineFormat at 0x...>
+    >>> cell.borders.top.width
+    0
+    >>> cell.borders.top.width = Pt(1)
+    >>> cell.borders.top.width
+    12700
+    >>> cell.borders.top.color.rgb = RGBColor(0, 0, 0)
+
+**Apply different line styles** Cell border work with ``LineFormat``, so any other line style can also be applied::
+
+    >>> cell.borders.top.dash_style = MSO_LINE_DASH_STYLE.DASH_DOT_DOT
```

Comment:
Okay, so I took a look at the line-format-parent interface as I mentioned earlier, and worked out a mechanism that no longer requires callbacks on the proxy objects and is flexible enough to accommodate this `_Cell` case that has six different lines on one object.

I have an implementation of it on my `develop` branch and will probably push it here shortly, but the gist of it is that the `LineFormat` constructor takes an ancestor of the line element (parent or grandparent) and the "name" of the line element. For example here the top border would be `LineFormat(tcPr, 'lnT')`. The ancestor element provided needs to implement `.get_or_add_lnT()` and `.lnT`. This aspect is similar to before but is now moved down into the oxml layer where those methods can often be shared by multiple proxy objects and in some cases will already be present (such as in this table cell case).

In reviewing your proposed API, I don't see the use in having a `_CellBorders` object. `_Cell` can have `.border_top`, `.border_left` etc. as direct properties, similar to how margin works. Each of these implementations is pretty trivial, something like this:

```python
class _Cell(..)
    ...
    @lazyproperty
    def border_bottom(self):
        """|LineFormat| object providing access to cell bottom border line style."""
        return LineFormat(self._tc.tcPr, 'lnB')
```

So if you want to rework this analysis document to reflect that API and implementation strategy we can get it committed.

A word on the PRs. I'll need them to be a groomed series of commits. This means no merges and no rework (like new commit to change or fix earlier work). You'll need to rebase in all your changes; otherwise it makes it harder for me to piece your PR into a series of commits I can add to master than it would be for me to do it over. Rebasing is very similar to merging as far as the skills involved, and allows you to produce groomed results that are easy to review and easy to commit from the bottom up, commit-by-commit, which is how we do it on this project. That way you don't have to get it all done or all right to contribute; if your next commit is acceptable we commit it on our feature branch and you can work on the rest from there.

Let me know if you need more explanation of what I mean on this.